File: BAEKJOON/java/17845_수강_과목.java
import java.io.*;
import java.util.StringTokenizer;

public class Main {
    public static int n, k;
    public static int[] importance, time;

    public static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    public static BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));

    public static void main(String[] args) throws IOException {
        input();
        solution();

        bw.flush();
        bw.close();
    }

    private static void input() throws IOException {
        StringTokenizer st = new StringTokenizer(br.readLine());
        n = Integer.parseInt(st.nextToken());
        k = Integer.parseInt(st.nextToken());

        importance = new int[k + 1];
        time = new int[k + 1];

        for (int i = 1; i <= k; i++) {
            st = new StringTokenizer(br.readLine());

            importance[i] = Integer.parseInt(st.nextToken());
            time[i] = Integer.parseInt(st.nextToken());
        }
    }

    private static void solution() throws IOException {
        int[][] dp = new int[k + 1][n + 1];

        for (int i = 1; i <= k; i++) {
            for (int j = 1; j <= n; j++) {
                if (time[i] > j) {
                    dp[i][j] = dp[i - 1][j];
                } else {
                    dp[i][j] = Math.max(dp[i - 1][j], dp[i - 1][j - time[i]] + importance[i]);
                }
            }
        }

        bw.write(String.valueOf(dp[k][n]));
    }
}

File: BAEKJOON/cpp/17845_수강_과목.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

int n, k;
vector<int> importance, time;

void input() {
	cin >> n >> k;
	
	importance.assign(k + 1, 0);
	time.assign(k + 1, 0);

	for (int i = 1; i <= k; i++) {
		cin >> importance[i] >> time[i];
	}
}

void solution() {
	vector<vector<int>> dp(k + 1, vector<int>(n + 1, 0));

	for (int i = 1; i <= k; i++) {
		for (int j = 1; j <= n; j++) {
			if (time[i] > j) {
				dp[i][j] = dp[i - 1][j];
			}
			else {
				dp[i][j] = max(dp[i - 1][j], dp[i - 1][j - time[i]] + importance[i]);
			}
		}
	}

	cout << dp[k][n];
}

int main() {
	ios::sync_with_stdio(false);
	cin.tie(NULL);
	cout.tie(NULL);

	input();
	solution();

	return 0;
}